File: packages/core/src/canonical-history.ts
import { archiveEventRecord } from './storage-digest.js'
import {
  StorageIntegrityError,
  type DurableSeriesState,
  type EventQueryOptions,
  type TaskEvent,
} from './types.js'

export function mergeCanonicalHistory(
  durableEvents: readonly TaskEvent[],
  hotEvents: readonly TaskEvent[],
  durableSeriesState: readonly DurableSeriesState[],
): TaskEvent[] {
  const seriesState = new Map<string, DurableSeriesState>()
  for (const state of durableSeriesState) {
    validateSeriesState(state)
    const key = seriesKey(state.taskId, state.seriesId)
    if (seriesState.has(key)) {
      throw new StorageIntegrityError(
        `Duplicate durable series state for ${state.taskId}:${state.seriesId}`,
      )
    }
    seriesState.set(key, state)
  }

  const byIndex = new Map<number, TaskEvent>()
  const byId = new Map<string, TaskEvent>()
  const add = (event: TaskEvent): void => {
    const indexed = byIndex.get(event.index)
    const identified = byId.get(event.id)
    if (indexed) {
      if (indexed.id !== event.id) {
        throw new StorageIntegrityError(
          `Canonical history index ${event.index} has conflicting event identities`,
        )
      }
      if (!sameEvent(indexed, event)) {
        throw new StorageIntegrityError(
          `Canonical history event ${event.id} has conflicting content`,
        )
      }
      return
    }
    if (identified) {
      throw new StorageIntegrityError(
        `Canonical history event ${event.id} has conflicting index or content`,
      )
    }
    byIndex.set(event.index, event)
    byId.set(event.id, event)
  }

  for (const event of durableEvents) {
    const state = matchingSeriesState(event, seriesState)
    if (state) continue
    add(event)
  }
  for (const state of durableSeriesState) add(state.event)

  for (const event of hotEvents) {
    const state = matchingSeriesState(event, seriesState)
    if (state && event.index <= state.throughIndex) continue
    add(event)
  }

  return Array.from(byIndex.values()).sort((left, right) => left.index - right.index)
}

export function applyCanonicalHistoryQuery(
  events: readonly TaskEvent[],
  opts?: EventQueryOptions,
): TaskEvent[] {
  let start = 0
  const since = opts?.since
  if (since?.id) {
    const position = events.findIndex((event) => event.id === since.id)
    if (position >= 0) start = position + 1
  }

  let result = events.slice(start)
  if (!since?.id && since?.index !== undefined) {
    result = result.filter((event) => event.index > since.index!)
  } else if (
    !since?.id &&
    since?.index === undefined &&
    since?.timestamp !== undefined
  ) {
    result = result.filter((event) => event.timestamp > since.timestamp!)
  }
  if (opts?.limit !== undefined) result = result.slice(0, opts.limit)
  return result
}

export function resolveCanonicalSeriesLatest(
  durableState: DurableSeriesState,
  hotEvents: readonly TaskEvent[],
): TaskEvent {
  validateSeriesState(durableState)
  const tail = hotEvents
    .filter(
      (event) =>
        event.taskId === durableState.taskId &&
        event.seriesId === durableState.seriesId &&
        event.seriesMode === durableState.mode &&
        event.index > durableState.throughIndex,
    )
    .sort((left, right) => left.index - right.index)

  if (durableState.mode === 'latest') {
    return tail.at(-1) ?? durableState.event
  }

  const field = durableState.event.seriesAccField ?? 'delta'
  let accumulated = durableState.event
  for (const event of tail) {
    const previousData = jsonObject(accumulated.data)
    const nextData = jsonObject(event.data)
    if (
      previousData &&
      nextData &&
      typeof previousData[field] === 'string' &&
      typeof nextData[field] === 'string'
    ) {
      accumulated = {
        ...event,
        data: {
          ...nextData,
          [field]: previousData[field] + nextData[field],
        },
      }
    } else {
      accumulated = event
    }
  }
  return accumulated
}

function matchingSeriesState(
  event: TaskEvent,
  states: ReadonlyMap<string, DurableSeriesState>,
): DurableSeriesState | undefined {
  if (
    !event.seriesId ||
    (event.seriesMode !== 'latest' && event.seriesMode !== 'accumulate')
  ) {
    return undefined
  }
  const state = states.get(seriesKey(event.taskId, event.seriesId))
  if (!state) return undefined
  if (state.mode !== event.seriesMode) {
    throw new StorageIntegrityError(
      `Canonical history series mode conflicts for ${event.taskId}:${event.seriesId}`,
    )
  }
  return state
}

function validateSeriesState(state: DurableSeriesState): void {
  if (
    state.event.taskId !== state.taskId ||
    state.event.seriesId !== state.seriesId ||
    state.event.seriesMode !== state.mode ||
    state.event.index !== state.throughIndex
  ) {
    throw new StorageIntegrityError(
      `Durable series state is inconsistent for ${state.taskId}:${state.seriesId}`,
    )
  }
}

function sameEvent(left: TaskEvent, right: TaskEvent): boolean {
  return archiveEventRecord(left) === archiveEventRecord(right)
}

function seriesKey(taskId: string, seriesId: string): string {
  return `${taskId}\0${seriesId}`
}

function jsonObject(value: unknown): Record<string, unknown> | null {
  return value !== null && typeof value === 'object' && !Array.isArray(value)
    ? value as Record<string, unknown>
    : null
}

File: rust/taskcast-core/src/canonical_history.rs
use std::collections::{BTreeMap, HashMap};

use crate::types::{
    DurableSeriesState, EventQueryOptions, SeriesMode, StorageIntegrityError, TaskEvent,
};

pub fn merge_canonical_history(
    durable_events: &[TaskEvent],
    hot_events: &[TaskEvent],
    durable_series_state: &[DurableSeriesState],
) -> Result<Vec<TaskEvent>, StorageIntegrityError> {
    let mut states = HashMap::new();
    for state in durable_series_state {
        validate_series_state(state)?;
        let key = series_key(&state.task_id, &state.series_id);
        if states.insert(key, state).is_some() {
            return Err(StorageIntegrityError::new(format!(
                "Duplicate durable series state for {}:{}",
                state.task_id, state.series_id
            )));
        }
    }

    let mut by_index = BTreeMap::<u64, TaskEvent>::new();
    let mut by_id = HashMap::<String, TaskEvent>::new();
    let mut add = |event: &TaskEvent| -> Result<(), StorageIntegrityError> {
        if let Some(indexed) = by_index.get(&event.index) {
            if indexed.id != event.id {
                return Err(StorageIntegrityError::new(format!(
                    "Canonical history index {} has conflicting event identities",
                    event.index
                )));
            }
            if !same_event(indexed, event) {
                return Err(StorageIntegrityError::new(format!(
                    "Canonical history event {} has conflicting content",
                    event.id
                )));
            }
            return Ok(());
        }
        if by_id.contains_key(&event.id) {
            return Err(StorageIntegrityError::new(format!(
                "Canonical history event {} has conflicting index or content",
                event.id
            )));
        }
        by_index.insert(event.index, event.clone());
        by_id.insert(event.id.clone(), event.clone());
        Ok(())
    };

    for event in durable_events {
        if matching_series_state(event, &states)?.is_none() {
            add(event)?;
        }
    }
    for state in durable_series_state {
        add(&state.event)?;
    }
    for event in hot_events {
        if matching_series_state(event, &states)?
            .is_some_and(|state| event.index <= state.through_index)
        {
            continue;
        }
        add(event)?;
    }

    Ok(by_index.into_values().collect())
}

pub fn apply_canonical_history_query(
    events: &[TaskEvent],
    opts: Option<EventQueryOptions>,
) -> Vec<TaskEvent> {
    let mut start = 0;
    let since = opts.as_ref().and_then(|opts| opts.since.as_ref());
    if let Some(id) = since.and_then(|since| since.id.as_ref()) {
        if let Some(position) = events.iter().position(|event| &event.id == id) {
            start = position + 1;
        }
    }

    let mut result = events[start..].to_vec();
    if since.and_then(|since| since.id.as_ref()).is_none() {
        if let Some(index) = since.and_then(|since| since.index) {
            result.retain(|event| event.index > index);
        } else if let Some(timestamp) = since.and_then(|since| since.timestamp) {
            result.retain(|event| event.timestamp > timestamp);
        }
    }
    if let Some(limit) = opts.and_then(|opts| opts.limit) {
        result.truncate(limit as usize);
    }
    result
}

pub fn resolve_canonical_series_latest(
    durable_state: &DurableSeriesState,
    hot_events: &[TaskEvent],
) -> Result<TaskEvent, StorageIntegrityError> {
    validate_series_state(durable_state)?;
    let mut tail = hot_events
        .iter()
        .filter(|event| {
            event.task_id == durable_state.task_id
                && event.series_id.as_deref() == Some(durable_state.series_id.as_str())
                && event.series_mode.as_ref() == Some(&durable_state.mode)
                && event.index > durable_state.through_index
        })
        .cloned()
        .collect::<Vec<_>>();
    tail.sort_by_key(|event| event.index);

    if durable_state.mode == SeriesMode::Latest {
        return Ok(tail
            .into_iter()
            .last()
            .unwrap_or_else(|| durable_state.event.clone()));
    }

    let field = durable_state
        .event
        .series_acc_field
        .as_deref()
        .unwrap_or("delta")
        .to_string();
    let mut accumulated = durable_state.event.clone();
    for event in tail {
        let previous = accumulated
            .data
            .as_object()
            .and_then(|data| data.get(&field))
            .and_then(serde_json::Value::as_str);
        let current = event
            .data
            .as_object()
            .and_then(|data| data.get(&field))
            .and_then(serde_json::Value::as_str);
        if let (Some(previous), Some(current), Some(mut data)) =
            (previous, current, event.data.as_object().cloned())
        {
            data.insert(
                field.clone(),
                serde_json::Value::String(format!("{previous}{current}")),
            );
            accumulated = TaskEvent {
                data: serde_json::Value::Object(data),
                ..event
            };
        } else {
            accumulated = event;
        }
    }
    Ok(accumulated)
}

fn matching_series_state<'a>(
    event: &TaskEvent,
    states: &'a HashMap<String, &'a DurableSeriesState>,
) -> Result<Option<&'a DurableSeriesState>, StorageIntegrityError> {
    let Some(series_id) = event.series_id.as_ref() else {
        return Ok(None);
    };
    if !matches!(
        event.series_mode,
        Some(SeriesMode::Latest | SeriesMode::Accumulate)
    ) {
        return Ok(None);
    }
    let Some(state) = states.get(&series_key(&event.task_id, series_id)).copied() else {
        return Ok(None);
    };
    if event.series_mode.as_ref() != Some(&state.mode) {
        return Err(StorageIntegrityError::new(format!(
            "Canonical history series mode conflicts for {}:{}",
            event.task_id, series_id
        )));
    }
    Ok(Some(state))
}

fn validate_series_state(state: &DurableSeriesState) -> Result<(), StorageIntegrityError> {
    if state.event.task_id != state.task_id
        || state.event.series_id.as_deref() != Some(state.series_id.as_str())
        || state.event.series_mode.as_ref() != Some(&state.mode)
        || state.event.index != state.through_index
    {
        return Err(StorageIntegrityError::new(format!(
            "Durable series state is inconsistent for {}:{}",
            state.task_id, state.series_id
        )));
    }
    Ok(())
}

fn same_event(left: &TaskEvent, right: &TaskEvent) -> bool {
    left.id == right.id
        && left.task_id == right.task_id
        && left.index == right.index
        && left.timestamp == right.timestamp
        && left.r#type == right.r#type
        && left.level == right.level
        && left.data == right.data
        && left.series_id == right.series_id
        && left.series_mode == right.series_mode
        && left.series_acc_field == right.series_acc_field
}

fn series_key(task_id: &str, series_id: &str) -> String {
    format!("{task_id}\0{series_id}")
}
